### DISCUSSION TITLE
dislay axis, bfl and image position, change surface

### DISCUSSION THREAD

**Discussion by ppfromero** [mjhoptics/ray-optics#85] [Q&A] [answered]
Hello guys,

I'm trying to compare 3 lenses performances:  My design / Aspherics / Doublet

I have some newbie question (sorry for this):
1. How to add axis to my layout figure (InteractiveLayout)
2. In my imported zmx lens, Image position is not located at BFL this is normal?
4. How to modify surface thickness after add surface?

Here below my script code

```
from rayoptics.environment import *

#-------------------------------------------------------------------
# My design

# Define optical model from scratch
opm = OpticalModel()
sm = opm['seq_model']
osp = opm['optical_spec']
pm = opm['parax_model']
em = opm['ele_model']
pt = opm['part_tree']


# System specs
opm.system_spec.title = 'My design'
opm.system_spec.dimensions = 'MM'
opm.radius_mode = True
sm.gaps[0].thi=1e10
osp['pupil'] = PupilSpec(osp, key=['object', 'pupil'], value=5)
osp['fov'] = FieldSpec(osp, key=['object', 'angle'], value=5, flds=[0., 1.0], is_relative=True)
osp['wvls'] = WvlSpec([(660, 1.0)], ref_wl=0)

# Build lens
sm.do_apertures = False
sm.add_surface([0., 10.,'','',6.25])
sm.set_stop()
sm.add_surface([49.57, 7.0, 'H-ZF10', 'CDGM',10])
sm.add_surface([-14.41, 1.97, 'H-LAK53B', 'CDGM',10])
sm.add_surface([-44.94, 37.49,'','',10])

# Update
opm.update_model()
#opm.save_model("My_design")

# Display properties
sm.list_model()
pm.first_order_data()

# Display Layout and spot
layout_plt = plt.figure(FigureClass=InteractiveLayout, opt_model=opm).plot()
spot_plt = plt.figure(FigureClass=SpotDiagramFigure, opt_model=opm,scale_type=Fit.All_Same, dpi=200).plot()

#-------------------------------------------------------------------
# Doublet

# Open zmx edmund doublet file
opm1 = open_model('doublet_zmax_32321.zmx') 
sm1 = opm1['seq_model']
osp1 = opm1['optical_spec']
pm1 = opm1['parax_model']
em1 = opm1['ele_model']
pt1 = opm1['part_tree']


# System specs
opm1.system_spec.title = 'Doublet Edmund 32321'
opm1.system_spec.dimensions = 'MM'
opm1.radius_mode = True
osp1['pupil'] = PupilSpec(osp1, key=['object', 'pupil'], value=5)
osp1['fov'] = FieldSpec(osp1, key=['object', 'angle'], value=5, flds=[0.,0.7], is_relative=True)
osp1['wvls'] = WvlSpec([(660, 1.0)], ref_wl=0)

# Update
opm1.update_model()

# Display properties
sm1.list_model()
pm1.first_order_data()

# Display Layout and spot
layout_plt1 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm1).plot()
spot_plt1 = plt.figure(FigureClass=SpotDiagramFigure, opt_model=opm1,scale_type=Fit.All_Same, dpi=200).plot()


#-------------------------------------------------------------------
# Aspherics

# Open zmx edmund aspherical file
opm2 = open_model('Aspherique_zmax_49664.zmx') 
sm2 = opm2['seq_model']
osp2 = opm2['optical_spec']
pm2 = opm2['parax_model']
em2 = opm2['ele_model']
pt2 = opm2['part_tree']


# System specs
opm2.system_spec.title = 'Aspheric Edmund 49664'
opm2.system_spec.dimensions = 'MM'
opm2.radius_mode = True
osp2['pupil'] = PupilSpec(osp2, key=['object', 'pupil'], value=5)
osp2['fov'] = FieldSpec(osp2, key=['object', 'angle'], value=5, flds=[0.,0.7], is_relative=True)
osp2['wvls'] = WvlSpec([(660, 1.0)], ref_wl=0)

# Update
opm2.update_model()

# Display properties
sm2.list_model()
pm2.first_order_data()

# Display Layout and spot
layout_plt2 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm2,is_dark=False).plot()
spot_plt2 = plt.figure(FigureClass=SpotDiagramFigure, opt_model=opm2,scale_type=Fit.All_Same, dpi=200).plot()

```

Thanks for your comments

**Comment by mjhoptics** [comment, marked as answer]
To answer your questions in reverse:

3. How to modify surface thickness after add surface? 

The sequential model, sm, has 2 lists, one for interfaces/surfaces and one for gaps, i.e. the thickness separating the adjacent interfaces and the material. The last gap in the list is the image distance,

`sm.gaps[-1]`

To change the image distance, do the following:

```
sm.gaps[-1].thi = 33.26
opm.update_model()
```

You may change more than one attribute before calling update_model().

2. In my imported zmx lens, Image position is not located at BFL this is normal?

The bfl specified on the webpage is different from the bfl (image distance) in the zmx file. This is an inconsistency in Edmund's data. It appears that the website bfl gives a smaller rms spot size across the field than the value in the zmx file.

1. How to add axis to my layout figure (InteractiveLayout)

I don't understand the question. Do you want a line drawn along the optical axis or somewhere else?

**Reply by ppfromero** [reply]
Thanks for your reply!

For question 1, when I use rayoptics GUI I have the following image with axis:
![image](https://user-images.githubusercontent.com/41118445/166416523-1d9eb449-67b0-4ea9-958e-5e43f1428f12.png)
When I use the following line:
`layout_plt2 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm2,is_dark=False).plot()`
The axis are not displayed...

Thanks

**Reply by mjhoptics** [reply]
Ah, *those* axes!

Use the do_draw_frame keyword arg when creating the layout. This sets an attribute of the same name, which you can change later if you wish.

`layout_plt1 = plt.figure(FigureClass=InteractiveLayout, opt_model=opm, is_dark=False, do_draw_frame=True).plot()`

**Reply by ppfromero** [reply]
Thanks for your help Michael!
